### PR TITLE
Add Support for Batch STARKs with Proving, Verification, and Recursion

### PR DESCRIPTION
**Pull request by sai-deng**
The method used in this PR can be found in Section 3.2 in the [paper](https://www.overleaf.com/read/gxqmchkvdrgp).

### CODE REVIEW

**Comment by Nashtare** [comment]
Hi @sai-deng! This is quite a consistent PR, would you mind posting a description of the changes, including the rationale please? Thank you!

**Comment by Nashtare** [comment]
Hey @sai-deng! Thank you for the PR! As this is quite dense, could you write some explanation about the approach taken, and any specific parts that may help reviewers?
Also could you explain how this impacts STARK recursive verification now that they're batched together? Thanks!

**Comment by sai-deng** [comment]
> Hey @sai-deng! Thank you for the PR! As this is quite dense, could you write some explanation about the approach taken, and any specific parts that may help reviewers? Also could you explain how this impacts STARK recursive verification now that they're batched together? Thanks!

The paper is still [WIP](https://www.overleaf.com/read/gxqmchkvdrgp),  the method used in this PR can be found in Section 3.2

File: plonky2/src/hash/field_merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    pub fn new(mut leaves: Vec<Vec<Vec<F>>>, cap_height: usize) -> Self {
+        assert!(!leaves.is_empty());
+        assert!(leaves.iter().all(|leaf| leaf.len().is_power_of_two()));
+        assert!(leaves
+            .windows(2)
+            .all(|pair| { pair[0].len() > pair[1].len() }));
+
+        let leaves_len = leaves[0].len();
+        let log2_leaves_len = log2_strict(leaves_len);
+        assert!(
```
Comment (hratoanina) [inline plonky2/src/hash/field_merkle_tree.rs:45]: I know it's not in the original code either, but maybe assert that the last log2 is >= cap _height?

File: plonky2/src/hash/field_merkle_tree.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            digests,
+            cap: MerkleCap(cap),
+            leaf_heights,
+        }
+    }
+
+    /// Create a Merkle proof from a leaf index.
+    pub fn open_batch(&self, leaf_index: usize) -> MerkleProof<F, H> {
+        let mut digests_buf_pos = 0;
+        let leaves_cap_height = log2_strict(self.leaves[0].len());
```
Comment (hratoanina) [inline plonky2/src/hash/field_merkle_tree.rs:134]: Extremely nit, but it confused me for a bit since the base layer isn't a cap, unlike all other layers.

**Review by hratoanina — COMMENTED** [review]
Still in the middle of reviewing. Looks really nice!

I have a random question: from what I'm seeing you can only open leaves from the bottom layer, and you open all intermediary leaves you encounter on your way to the top. If for some reason you want to open only a leaf from one of the smaller trees, can't you have a shorter proof starting at the smaller tree's height?

**Comment by sai-deng** [comment]
> Still in the middle of reviewing. Looks really nice!
> 
> I have a random question: from what I'm seeing you can only open leaves from the bottom layer, and you open all intermediary leaves you encounter on your way to the top. If for some reason you want to open only a leaf from one of the smaller trees, can't you have a shorter proof starting at the smaller tree's height?

Thanks for reviewing! In this case, you'd better not use a field Merkle tree since it will not reduce the Merkle proof size.

File: plonky2/src/hash/field_merkle_tree.rs (hunk 2 above)
Comment (sai-deng) [reply plonky2/src/hash/field_merkle_tree.rs]: Done.

File: plonky2/src/hash/field_merkle_tree.rs (hunk 1 above)
Comment (sai-deng) [reply plonky2/src/hash/field_merkle_tree.rs:44]: Yes! We should compare last log2.

File: plonky2/src/hash/field_merkle_tree.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
+    /// Each element in the `leaves` vector represents a matrix (a vector of vectors).
+    /// The height of each matrix should be a power of two.
+    /// The `leaves` vector should be sorted by matrix height, from tallest to shortest, with no duplicate heights.
+    // TODO: FieldMerkleTree does not handle duplicates; this is deferred to the caller.  Revisit when implementing batch FRI to potentially optimize.
```
Comment (Nashtare) [inline plonky2/src/hash/field_merkle_tree.rs:35]: nit: is this TODO to remain?

**Review by Nashtare — COMMENTED** [review]
Fairly minor comments as well for now, will do another pass at FRI later.

File: plonky2/src/hash/field_merkle_tree.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+use crate::hash::hash_types::{RichField, NUM_HASH_OUT_ELTS};
+use crate::hash::merkle_proofs::MerkleProof;
+use crate::hash::merkle_tree::{
+    capacity_up_to_mut, fill_digests_buf, merkle_tree_prove, MerkleCap,
+};
+use crate::plonk::config::{GenericHashOut, Hasher};
+use crate::util::log2_strict;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/hash/field_merkle_tree.rs]: I agree with it. I was using the [name](https://github.com/Plonky3/Plonky3/blob/main/merkle-tree/src/merkle_tree.rs) from Plonky3. Let's rename it to BatchMerkleTree.

File: plonky2/src/hash/merkle_proofs.rs [hunk 5]
```diff
@@ -76,6 +76,45 @@ pub fn verify_merkle_proof_to_cap<F: RichField, H: Hasher<F>>(
...
+        };
+        current_height -= 1;
+
+        if leaf_data_index < leaf_heights.len() && current_height == leaf_heights[leaf_data_index] {
+            let mut new_leaves = current_digest.to_vec();
+            new_leaves.extend_from_slice(&leaf_data[leaf_data_index]);
+            current_digest = H::hash_or_noop(&new_leaves);
+            leaf_data_index += 1;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/hash/merkle_proofs.rs]: Done.

File: plonky2/src/fri/batch_oracle.rs [hunk 6]
```diff
@@ -0,0 +1,610 @@
+#[cfg(not(feature = "std"))]
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/fri/batch_oracle.rs]: Moved files to src/batch_fri

File: plonky2/src/fri/validate_shape.rs [hunk 7]
```diff
@@ -65,3 +68,64 @@ where
 
     Ok(())
 }
+
+pub(crate) fn validate_batch_fri_proof_shape<F, C, const D: usize>(
+    proof: &FriProof<F, C::Hasher, D>,
+    instances: &[FriInstanceInfo<F, D>],
+    params: &FriParams,
+) -> anyhow::Result<()>
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/fri/validate_shape.rs:30]: Done.

File: plonky2/src/hash/field_merkle_tree.rs (hunk 3 above)
Comment (sai-deng) [reply plonky2/src/hash/field_merkle_tree.rs]: Removed

File: plonky2/src/hash/merkle_tree.rs [hunk 8]
```diff
@@ -148,6 +148,47 @@ fn fill_digests_buf<F: RichField, H: Hasher<F>>(
     );
 }
 
+pub fn merkle_tree_prove<F: RichField, H: Hasher<F>>(
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/hash/merkle_tree.rs]: It should not be publicly exposed. Fixed.

**Comment by sai-deng** [comment]
> Fairly minor comments as well for now, will do another pass at FRI later.

Thanks for the careful reviews! Comments are addressed in [59d637f](https://github.com/0xPolygonZero/plonky2/pull/1600/commits/59d637ff8fad34abc68839762563cd1737961a31).

File: plonky2/src/batch_fri/oracle.rs [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+            polynomials,
+            batch_merkle_tree,
+            degree_bits,
+            rate_bits,
+            blinding,
+        }
+    }
+
+    /// Produces a batch opening proof.
+    pub fn prove_openings(
```
Comment (hratoanina) [inline plonky2/src/batch_fri/oracle.rs:128]: If I understand correctly the logic:

You provide several `FriInstanceInfo`, which consist in multiple `FriBatchInfo`, containing each an evaluation point and a batch of polynomials to be evaluated at that point. In the batched version, for a single `FriInstanceInfo`, the evaluated polynomials can be of different degrees, contrary to the normal version.
But I don't think the inner `for` loop (at line 154) works well with polynomials of different degrees. To my understanding, what we should be doing is to group all polynomials of the same degree _across all instances_, and then compose them together with `alpha`.

**Review by hratoanina — COMMENTED** [review]
Still reviewing, thanks!

**Comment by hratoanina** [comment]
I have a question with regards to STARK batching in general. From section 3.3 of the paper, I understand that for each individual STARK you pad it to the maximum length it can have given a normal execution (determined via testing with a bunch of different inputs) to have a fixed length you can build a recursion circuit from.
In our zkEVM, we have very wide discrepancies between best and worst case scenarios for individual trace lengths, which can range e.g. from 16 to 2^20. Does that mean we would be forced to make a circuit accommodating a 2^20-length trace, and pad the trace to this length for every single execution?

**Comment by sai-deng** [comment]
> I have a question with regards to STARK batching in general. From section 3.3 of the paper, I understand that for each individual STARK you pad it to the maximum length it can have given a normal execution (determined via testing with a bunch of different inputs) to have a fixed length you can build a recursion circuit from. In our zkEVM, we have very wide discrepancies between best and worst case scenarios for individual trace lengths, which can range e.g. from 16 to 2^20. Does that mean we would be forced to make a circuit accommodating a 2^20-length trace, and pad the trace to this length for every single execution?

Yes, we have to pad the trace to a length of 2^20. Alternatively, we can use a Merkle tree to save all the valid recursion circuits hash and treat the Merkle root hash as the public parameters.

File: plonky2/src/batch_fri/oracle.rs [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+use crate::plonk::config::GenericConfig;
+use crate::timed;
+use crate::util::reducing::ReducingFactor;
+use crate::util::timing::TimingTree;
+use crate::util::{reverse_bits, transpose};
+
+/// Represents a batch FRI oracle, i.e. a batch of polynomials with different degrees which have
+/// been Merkle-ized in a [`BatchMerkleTree`].
+#[derive(Eq, PartialEq, Debug)]
+pub struct BatchFriOracle<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/batch_fri/oracle.rs:30]: Yes, I think we should rename the later FriOracle but let us do it in a separate PR.

File: plonky2/src/batch_fri/oracle.rs (hunk 9 above)
Comment (sai-deng) [reply plonky2/src/batch_fri/oracle.rs:128]: In the batched version, for a single FriInstanceInfo, the evaluated polynomials have the same degree.

Comment (sai-deng) [reply plonky2/src/batch_fri/oracle.rs:128]: In zkVM code, we first create a FriInstanceInfo for each table. We then merge FriInstanceInfo instances when two tables have the same degree. Lastly, we have a vec<FriInstanceInfo> with different degrees, but within each FriInstanceInfo, the polynomials have the same degree.

File: plonky2/src/hash/merkle_proofs.rs [hunk 11]
```diff
@@ -151,6 +190,62 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
...
+                elements: hash_outs,
+            };
+            current_height -= 1;
+
+            if leaf_data_index < leaf_heights.len()
+                && current_height == leaf_heights[leaf_data_index]
+            {
+                let mut new_leaves = state.elements.to_vec();
+                new_leaves.extend_from_slice(&leaf_data[leaf_data_index]);
+                state = self.hash_or_noop::<H>(new_leaves.clone());
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/hash/merkle_proofs.rs:224]: Done.

File: plonky2/src/batch_fri/prover.rs [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+        final_coeffs = PolynomialCoeffs::new(
+            final_coeffs
+                .coeffs
+                .par_chunks_exact(arity)
+                .map(|chunk| reduce_with_powers(chunk, beta))
+                .collect::<Vec<_>>(),
+        );
+        shift = shift.exp_u64(arity as u64);
+        final_values = final_coeffs.coset_fft(shift.into());
+        if polynomial_index != values.len() && final_values.len() == values[polynomial_index].len()
```
[the remark this answers is not in the document]
Comment (sai-deng) [reply plonky2/src/batch_fri/prover.rs:112]: Done.

**Comment by hratoanina** [comment]
Hi @sai-deng, sorry for the wait!
I'm currently implementing STARK batching on our zkEVM alongside the review, which I treat as another layer of testing. I should be done by the end of the week; if there's no issue I'll approve the PR then!

**Comment by sai-deng** [comment]
> Hi @sai-deng, sorry for the wait! I'm currently implementing STARK batching on our zkEVM alongside the review, which I treat as another layer of testing. I should be done by the end of the week; if there's no issue I'll approve the PR then!

Batch STARKs are already implemented in our zkVM; unfortunately, the code is not open-sourced yet. Let me know if you need any help during the implementation.

**Review by Nashtare — APPROVED** [review]
Thanks again!

**Pull request merged**